cmd/nodeJsBuild.go

```diff
@@ -0,0 +1,161 @@
+	fileUtils  FileUtils.Files
+	execRunner *command.Command
+}
+
+func (u *nodeJsBuildUtilsBundle) fileExists(path string) (bool, error) {
+	return u.fileUtils.FileExists(path)
+}
+
+func (u *nodeJsBuildUtilsBundle) glob(pattern string) (matches []string, err error) {
+	return doublestar.Glob(pattern)
+}
+
+func (u *nodeJsBuildUtilsBundle) getwd() (dir string, err error) {
+	return os.Getwd()
+}
+
+func (u *nodeJsBuildUtilsBundle) chdir(dir string) error {
+	return os.Chdir(dir)
+}
+
+func (u *nodeJsBuildUtilsBundle) getExecRunner() execRunner {
+	if u.execRunner == nil {
+		u.execRunner = &command.Command{}
+		u.execRunner.Stdout(log.Entry().Writer())
+		u.execRunner.Stderr(log.Entry().Writer())
+	}
+	return u.execRunner
+}
+
+func nodeJsBuild(config nodeJsBuildOptions, telemetryData *telemetry.CustomData) {
+	utils := nodeJsBuildUtilsBundle{}
+
+	err := runNodeJsBuild(&utils, &config)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+func runNodeJsBuild(utils nodeJsBuildUtilsInterface, options *nodeJsBuildOptions) error {
+	execRunner := utils.getExecRunner()
+	log.Entry().Infof("NPM Registry configuration: defaultNpmRegistry %s, sapNpmRegistry %s",
+		options.DefaultNpmRegistry, options.SapNpmRegistry)
+	environment := []string{"npm_config_@sap:registry=" + options.SapNpmRegistry}
+	if options.DefaultNpmRegistry != "" {
+		environment = append(environment, "npm_config_registry="+options.DefaultNpmRegistry)
+	}
+	execRunner.SetEnv(environment)
+
+	packageJSONFiles, err := findPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+
+	for _, file := range packageJSONFiles {
+		dir := path.Dir(file)
+		err = utils.chdir(dir)
+		if err != nil {
+			return err
+		}
+		packageLockExists, err := utils.fileExists("package-lock.json")
+
+		if err != nil {
+			return err
+		}
+		yarnLockExists, err := utils.fileExists("yarn.lock")
+		if err != nil {
+			return err
+		}
+		if options.Install {
+			err = installDependencies(dir, packageLockExists, yarnLockExists, execRunner)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, v := range options.RunScripts {
+			log.Entry().WithField("WorkingDirectory", dir).Info("run-script " + v)
+			err = execRunner.RunExecutable("npm", "run-script", v, "--if-present")
+			if err != nil {
+				return err
+			}
+		}
+		err = utils.chdir(oldWorkingDirectory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func findPackageJSONFiles(utils nodeJsBuildUtilsInterface) ([]string, error) {
+	unfilteredListOfPackageJSONFiles, err := utils.glob("**/package.json")
+	if err != nil {
+		return nil, err
+	}
+
+	var packageJSONFiles []string
+
+	for _, file := range unfilteredListOfPackageJSONFiles {
+		if strings.Contains(file, "node_modules") {
+			continue
+		}
+		projectStructure := FileUtils.ProjectStructure{}
+		isMta := projectStructure.UsesMta()
+		if isMta && strings.Contains(file, "gen/") {
```

Comment:
This check seems too fragile to me. The value of `file` could be "somewhere/hallucinogen/whatever/package.json" and you would filter it out.